### PR TITLE
v1.1.0: Private and message-based state catch-up, Sync namespace, Session.js, Connection Info, Socket Repair Center

### LINKED ISSUE #12: better logging
- anytime we log anything, include the serialized session state

### LINKED ISSUE #15: Handle network disconnect reasons nicely
Working on this already. 

Try to reconnect on ping timeouts; boot clients for any other reason.

### PR DESCRIPTION
**Pull request by parseccentric**
# v1.1.0

**Private and message-based state catch-up**

Now state catch-up is only sent to the client who requested it. Previously, it would get sent to every client, including the one who requested it.

**Sync Namespace**

Breaking change: all clients must connect to "sync" namespace instead of just the default namespace.

Example client code:

`var socket = io(window.RELAY_BASE_URL + "/sync"); //CURRENT WAY`
`var socket = io(window.RELAY_BASE_URL); //DEPRECATED, OLD WAY`

The motivation for this is that every socket that connects to the "/chat" and "/admin" namespaces would also connect to the default "/" namespace, making it difficult to tell whether a sync connection was intended and broken, or if it was a chat or admin connection. This also makes it easier to debug and log information about the sockets.

**Session.js**

Moved many functions from sync.js into session.js for better object-oriented programming.

**Connection Info**

More events are available for clients. For example, they can request the server name, defined in a constant.

New session events are available to the clients: 
* Successfully joined
* Failed to join
* Successfully left
* Failed to leave

**Socket Repair Center**

Instead of getting automatically joined to a room/session, the sync server now places newly connected sockets to socket repair center.

The sync server looks at the message metadata to try and connect sockets that are connected but not joined to a session.

Invalid sockets get disconnected by the server.

When a socket is successfully repaired, the sync server sends a state catch-up event.

Fixes #15 

**Bug Fixes**

**Logging**

The white space in the logging now uses tabs rather than spaces.

The session state is logged every time a catch-up is requested. 

Fixes #12 

**Refactors**

Many things were renamed.

Constants were introduced to differentiate SocketIO event names from Komodo Send Events and Komodo Receive Events.

The process of opening a message and applying its action to the sync server's state object was refactored into a large amount of smaller methods for clarity. 

**Tests**

The unit tests were split into multiple files. 

* test-sync-clients-and-sockets.js
* test-sync-integration.js
* test-sync-sessions.js

Some tests for session.js were introduced.

* It should be able to bump one existing socket
* It should be able to remove a socket
* It should create a correct sockets object

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Testing 
- [x] Change that requires a documentation update

# How Has This Been Tested?

Run unit tests -> Unit tests pass.

Run two Unity clients with the same session ID and client ID -> the People tab indicates a bump in one window.

Then press Leave & Rejoin or Close Connection & Rejoin on one -> the People tab indicates successful connection in one window, and a bump in the other window.

Run two Unity clients in Chrome with one's network throttling set to Offline and press a reconnect button -> Connection error is displayed in People tab

Then set network throttling to normal and wait -> Connection is automatically restored and client movement is reflected in each

- [x] Manual Test
- [x] Unit Test
- [x] Integration / End-to-End Test

**Test Configuration**:

* Firefox v??
* Oculus Quest v??
- [x] PC VR (Link)
- [ ] Standalone 

# Checklist:

- [ ] ~~My code follows the style guidelines of this project~~
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas that are not self-documenting
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings 
- [x] My changes have no unnecessary logging
- [ ] I have added tests that prove my fix is effective or that my feature works, for sufficiently complex features
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] Sensitive info like tokens, secrets, and passwords have been removed before submitting

Modified from this article:
Phillip Johnston, “A GitHub Pull Request Template for Your Projects - Embedded Artistry,” Embedded Artistry, Aug. 04, 2017. https://embeddedartistry.com/blog/2017/08/04/a-github-pull-request-template-for-your-projects/ (accessed Jul. 22, 2021).

### CODE REVIEW

File: sync.js [hunk 1]
```diff
@@ -56,1683 +64,2855 @@ const { syslog } = require('winston/lib/winston/config');
...
-            } else if (session && !session.isRecording) {
-                if (this.logger) this.logger.warn(`Requested to end session capture, but capture is already ended: ${session_id}`);
-                session.capture_id = null;
-            } else {
-                if (this.logger) this.logger.warn(`Error ending capture for session: ${session_id}`);
-            }
-        }
-    },
+      // don't call failedToLeaveAction here because we don't have a socket.
+
```
Comment (Barasakar) [inline sync.js:1211]: Probably you don't need to space for line 1211.

**Review by Barasakar — COMMENTED** [review]
I am not sure if you want me to do Comment, Approve, or Request changes, but for now I will do comment. 
All of my suggestions are about formatting.

**Comment by parseccentric** [comment]
Thanks so much for these comments. I will take a look soon.

File: admin.js [hunk 2]
```diff
@@ -59,10 +61,11 @@ module.exports = {
         });
 
         admin.on('connection', function(socket) { //TODO finish or remove.
+            // TODO(Brandon): log connection here
+
             socket.emit("adminInfo", socket.id);
 
             socket.on('getAllSessions0', function() {
-
                 this.sessions = syncServer.sessions;
```
[the remark this answers is not in the document]
Comment (parseccentric) [reply admin.js:70]: Oh, that is a good idea. We should make that our style guide.

**Comment by parseccentric** [comment]
Thank you @Barasakar! When you are ready can you press "Merge pull request" and then let me know?

**Pull request merged**